File: Grille.java
import java.util.ArrayList;
import java.util.List;
import java.util.HashMap; // Option pour stocker les cellules si la grille est sparse
                          // Mais pour une grille dense, un tableau 2D de Listes est souvent plus direct.

public class Grille {
    public final int tailleX;
    public final int tailleY;

    // Structure pour stocker les agents sur la grille.
    // Un tableau 2D où chaque élément est une liste d'agents présents sur cette cellule.
    // List<Agent>[][] cellules; // Syntaxe pour un tableau de Listes d'Agents
    // Ou, pour une initialisation plus simple et flexible avec ArrayList:
    private List<Agent>[][] cellules;


    @SuppressWarnings("unchecked") // Pour la création du tableau de listes génériques
    public Grille(int tailleX, int tailleY) {
        if (tailleX <= 0 || tailleY <= 0) {
            throw new IllegalArgumentException("Les dimensions de la grille doivent être positives.");
        }
        this.tailleX = tailleX;
        this.tailleY = tailleY;

        // Initialisation du tableau de listes
        // On ne peut pas créer un tableau de génériques directement comme new List<Agent>[tailleX][tailleY];
        // Il faut passer par un cast depuis un tableau d'Object ou un tableau de List non typée.
        this.cellules = (ArrayList<Agent>[][]) new ArrayList[tailleX][tailleY];
        for (int i = 0; i < tailleX; i++) {
            for (int j = 0; j < tailleY; j++) {
                this.cellules[i][j] = new ArrayList<Agent>();
            }
        }
    }

    /**
     * Réinitialise la grille en vidant toutes les listes d'agents des cellules.
     */
    public void initialiserGrille() {
        for (int i = 0; i < tailleX; i++) {
            for (int j = 0; j < tailleY; j++) {
                this.cellules[i][j].clear();
            }
        }
    }

    /**
     * Ajoute un agent à sa position actuelle sur la grille.
     * @param agent L'agent à ajouter.
     */
    public void ajouterAgent(Agent agent) {
        if (agent == null) return;
        if (agent.posX >= 0 && agent.posX < tailleX && agent.posY >= 0 && agent.posY < tailleY) {
            this.cellules[agent.posX][agent.posY].add(agent);
        } else {
            // Gérer l'erreur: agent hors limites.
            // Pourrait lancer une exception ou afficher un message.
            System.err.println("Tentative d'ajout de l'agent " + agent.agentId + " à des coordonnées hors limites : (" + agent.posX + "," + agent.posY + ")");
        }
    }

    /**
     * Retire un agent de sa position spécifiée (oldX, oldY) sur la grille.
     * @param agent L'agent à retirer.
     * @param oldX Ancienne coordonnée X de l'agent.
     * @param oldY Ancienne coordonnée Y de l'agent.
     */
    public void retirerAgent(Agent agent, int oldX, int oldY) {
        if (agent == null) return;
        if (oldX >= 0 && oldX < tailleX && oldY >= 0 && oldY < tailleY) {
            this.cellules[oldX][oldY].remove(agent); // remove() sur une List utilise equals() pour trouver l'objet.
                                                   // S'assure que Agent a une implémentation correcte de equals() si nécessaire,
                                                   // mais ici on retire l'instance exacte.
        } else {
             System.err.println("Tentative de retrait de l'agent " + agent.agentId + " de coordonnées hors limites : (" + oldX + "," + oldY + ")");
        }
    }

    /**
     * Met à jour la position d'un agent sur la grille.
     * L'agent doit déjà avoir ses nouvelles coordonnées (posX, posY) mises à jour.
     * @param agent L'agent dont la position est mise à jour.
     * @param oldX Ancienne coordonnée X de l'agent.
     * @param oldY Ancienne coordonnée Y de l'agent.
     */
    public void mettreAJourPositionAgent(Agent agent, int oldX, int oldY) {
        retirerAgent(agent, oldX, oldY);
        ajouterAgent(agent); // L'agent a déjà ses nouvelles coordonnées
    }

    /**
     * Retourne la liste des agents présents sur une cellule spécifique.
     * @param x Coordonnée X de la cellule.
     * @param y Coordonnée Y de la cellule.
     * @return Une liste d'agents. Peut être vide si aucun agent n'est sur la cellule.
     */
    public List<Agent> getAgentsSurCellule(int x, int y) {
        // L'appelant est responsable de la gestion de la toroidalité pour les coordonnées x, y
        // avant d'appeler cette méthode si un accès toroïdal est souhaité.
        if (x >= 0 && x < tailleX && y >= 0 && y < tailleY) {
            return this.cellules[x][y];
        } else {
            // Coordonnées hors des limites brutes de la grille
            System.err.println("Accès hors limites dans getAgentsSurCellule pour (" + x + "," + y + "). Retourne une liste vide.");
            return new ArrayList<Agent>(); // Retourne une liste vide pour éviter NullPointerException
        }
    }

    // La méthode get_voisinage_moore n'est pas strictement nécessaire ici si la logique
    // de voisinage est entièrement gérée dans Agent.verifierInfection, comme c'est le cas.
    // Si elle était nécessaire, elle ressemblerait à ceci :
    /*
    public List<Agent> getVoisinageMoore(int xCentre, int yCentre) {
        List<Agent> voisinsComplets = new ArrayList<>();
        if (xCentre < 0 || xCentre >= tailleX || yCentre < 0 || yCentre >= tailleY) {
            System.err.println("Coordonnées centrales hors limites pour getVoisinageMoore.");
            return voisinsComplets;
        }

        for (int dx = -1; dx <= 1; ++dx) {
            for (int dy = -1; dy <= 1; ++dy) {
                int voisinX = (xCentre + dx + tailleX) % tailleX;
                int voisinY = (yCentre + dy + tailleY) % tailleY;
                voisinsComplets.addAll(getAgentsSurCellule(voisinX, voisinY));
            }
        }
        return voisinsComplets;
    }
    */
}

File: Grille.cpp
// Grille.cpp
#include "Grille.hpp"
#include <stdexcept> // Pour std::invalid_argument ou std::out_of_range
#include <iostream>  // Pour d'éventuels messages d'erreur (std::cerr)

Grille::Grille(int tx, int ty) : taille_x(tx), taille_y(ty) {
    if (tx <= 0 || ty <= 0) {
        // Il est préférable de lancer une exception si les dimensions ne sont pas valides.
        throw std::invalid_argument("Les dimensions de la grille (taille_x, taille_y) doivent être strictement positives.");
    }
    // Redimensionne le vecteur externe (correspondant à la dimension x)
    cellules.resize(taille_x);
    for (int i = 0; i < taille_x; ++i) {
        // Redimensionne chaque vecteur interne (correspondant à la dimension y)
        cellules[i].resize(taille_y);
        // À ce stade, cellules[i][j] est un std::vector<Agent*> vide.
    }
}

void Grille::initialiser_grille() {
    // Vide la liste d'agents pour chaque cellule de la grille.
    // Utile si on veut réutiliser l'objet Grille pour une nouvelle simulation sans le reconstruire.
    for (int i = 0; i < taille_x; ++i) {
        for (int j = 0; j < taille_y; ++j) {
            cellules[i][j].clear();
        }
    }
}

// Ajoute un agent à sa position actuelle (déjà définie dans l'objet agent)
void Grille::ajouter_agent(Agent* agent) {
    if (!agent) {
        // Optionnel : gérer le cas d'un pointeur nul, bien que cela ne devrait pas arriver si bien utilisé.
        // std::cerr << "Tentative d'ajout d'un agent nul à la grille." << std::endl;
        return;
    }
    // Vérifier si les coordonnées de l'agent sont dans les limites de la grille
    if (agent->pos_x >= 0 && agent->pos_x < taille_x &&
        agent->pos_y >= 0 && agent->pos_y < taille_y) {
        cellules[agent->pos_x][agent->pos_y].push_back(agent);
    } else {
        // Gérer l'erreur : agent avec des coordonnées hors limites.
        // Ceci pourrait indiquer un problème logique ailleurs.
        // throw std::out_of_range("Tentative d'ajout d'un agent à des coordonnées hors limites.");
        // std::cerr << "Erreur: Agent " << agent->agent_id << " hors limites (" << agent->pos_x << "," << agent->pos_y << ") pour grille " << taille_x << "x" << taille_y << std::endl;
    }
}

// Retire un agent de la cellule spécifiée par old_x, old_y
void Grille::retirer_agent(Agent* agent, int old_x, int old_y) {
    if (!agent) return;

    // Vérifier si les anciennes coordonnées sont valides
    if (old_x >= 0 && old_x < taille_x &&
        old_y >= 0 && old_y < taille_y) {

        auto& agents_dans_cellule = cellules[old_x][old_y];
        // Recherche et suppression de l'agent dans le vecteur de la cellule
        for (size_t i = 0; i < agents_dans_cellule.size(); ++i) {
            if (agents_dans_cellule[i] == agent) {
                agents_dans_cellule.erase(agents_dans_cellule.begin() + i);
                break; // On suppose qu'un agent n'est listé qu'une fois par cellule
            }
        }
        // Optionnel : avertir si l'agent n'a pas été trouvé à old_pos,
        // cela pourrait indiquer une incohérence.
    } else {
        // Gérer l'erreur : anciennes coordonnées hors limites.
        // throw std::out_of_range("Tentative de retrait d'un agent de coordonnées hors limites.");
        // std::cerr << "Erreur: Coordonnées de retrait hors limites (" << old_x << "," << old_y << ") pour agent " << agent->agent_id << std::endl;
    }
}

// Met à jour la position de l'agent sur la grille.
// Prérequis : agent->pos_x et agent->pos_y contiennent déjà les NOUVELLES coordonnées.
void Grille::mettre_a_jour_position_agent(Agent* agent, int old_x, int old_y) {
    if (!agent) return;
    retirer_agent(agent, old_x, old_y); // Retire de l'ancienne position
    ajouter_agent(agent);              // Ajoute à la nouvelle position (stockée dans l'objet agent)
}

const std::vector<Agent*>& Grille::get_agents_sur_cellule(int x, int y) const {
    // Cette fonction attend des coordonnées DÉJÀ normalisées (par ex. après application de la toroidalité par l'appelant).
    // Si ce n'est pas le cas, et que x,y peuvent être hors limites brutes, il faut les normaliser ici ou lancer une erreur.
    // Pour la cohérence avec Agent::verifier_infection, l'appelant gère la toroidalité.
    if (x < 0 || x >= taille_x || y < 0 || y >= taille_y) {
        // Ceci indique une erreur logique si l'appelant était censé normaliser.
        // throw std::out_of_range("Accès à get_agents_sur_cellule avec des coordonnées brutes hors limites.");
        // Pour la robustesse, on peut retourner une référence à un vecteur vide statique.
        static const std::vector<Agent*> empty_vector;
        // std::cerr << "Avertissement : Coordonnées (" << x << "," << y << ") hors limites dans get_agents_sur_cellule. Retourne vecteur vide." << std::endl;
        return empty_vector;
    }
    return cellules[x][y];
}

// Renvoie une liste de tous les agents (pointeurs) dans le voisinage de Moore (les 8 cellules autour + la cellule centrale)
// en gérant la toroidalité.
std::vector<Agent*> Grille::get_voisinage_moore(int x_centre, int y_centre) const {
    std::vector<Agent*> agents_voisins;

    // Vérifier si les coordonnées du centre sont valides (optionnel, dépend si l'appelant garantit cela)
    if (x_centre < 0 || x_centre >= taille_x || y_centre < 0 || y_centre >= taille_y) {
        // std::cerr << "Avertissement : Coordonnées centrales hors limites pour get_voisinage_moore." << std::endl;
        return agents_voisins; // Retourne un vecteur vide
    }

    for (int dx = -1; dx <= 1; ++dx) {
        for (int dy = -1; dy <= 1; ++dy) {
            // Calcul des coordonnées du voisin avec gestion de la toroidalité
            int voisin_x = (x_centre + dx + taille_x) % taille_x;
            int voisin_y = (y_centre + dy + taille_y) % taille_y;

            // Récupérer les agents de la cellule voisine (déjà normalisée)
            const std::vector<Agent*>& agents_cellule_actuelle = get_agents_sur_cellule(voisin_x, voisin_y);
            // Ajouter tous les agents de cette cellule à la liste des voisins
            agents_voisins.insert(agents_voisins.end(), agents_cellule_actuelle.begin(), agents_cellule_actuelle.end());
        }
    }
    return agents_voisins;
}
